hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemAttributes.java

```diff
@@ -42,70 +49,194 @@ public ITestAzureBlobFileSystemAttributes() throws Exception {
+    final Path path = new Path("a/b");
+    fs.create(path);
+
+    String attributeName1 = "user.attribute1";
+    String attributeName2 = "user.attribute2";
+    String decodedAttributeValue1;
+    String decodedAttributeValue2;
+    byte[] attributeValue1;
+    byte[] attributeValue2;
+
+    if(fs.getAbfsStore().getPrefixMode() == PrefixMode.BLOB) {
+      Assume.assumeTrue(!getIsNamespaceEnabled(fs)); // Blob endpoint Currently Supports FNS only
+      decodedAttributeValue1 = "hi";
+      decodedAttributeValue2 = "Блюз"; //Блюз //你好
+      attributeValue1 = decodedAttributeValue1.getBytes(StandardCharsets.UTF_8);
+      attributeValue2 = decodedAttributeValue2.getBytes(StandardCharsets.UTF_8);
+    }
+    else {
+      decodedAttributeValue1 = "hi";
+      decodedAttributeValue2 = "hello"; // DFS Endpoint only Supports ASCII
+      attributeValue1 = fs.getAbfsStore().encodeAttribute(decodedAttributeValue1);
+      attributeValue2 = fs.getAbfsStore().encodeAttribute(decodedAttributeValue2);
+    }
+
+    // Attribute not present initially
+    assertNull(fs.getXAttr(path, attributeName1));
+    assertNull(fs.getXAttr(path, attributeName2));
+
+    // Set the Attributes
+    fs.setXAttr(path, attributeName1, attributeValue1);
+
+    // Check if the attribute is retrievable
+    byte[] rv = fs.getXAttr(path, attributeName1);
+    assertTrue(Arrays.equals(rv, attributeValue1));
+    assertEquals(new String(rv, StandardCharsets.UTF_8), decodedAttributeValue1);
+
+    // Set the second Attribute
+    fs.setXAttr(path, attributeName2, attributeValue2);
+
+    // Check all the attributes present and previous Attribute not overridden
+    rv = fs.getXAttr(path, attributeName1);
+    assertTrue(Arrays.equals(rv, attributeValue1));
+    assertEquals(new String(rv, StandardCharsets.UTF_8), decodedAttributeValue1);
+    rv = fs.getXAttr(path, attributeName2);
+    assertTrue(Arrays.equals(rv, attributeValue2));
+    assertEquals(new String(rv, StandardCharsets.UTF_8), decodedAttributeValue2);
+  }
+
+  /**
+   * Trying to set same attribute multiple times should result in no failure
+   * @throws Exception
+   */
+  @Test
+  public void testSetXAttrMultipleOperations() throws Exception {
+    AzureBlobFileSystem fs = getFileSystem();
+    final Path path = new Path("a/b");
+    fs.create(path);
+
+    String attributeName1 = "user.attribute1";
+    byte[] attributeValue1;
+    String decodedAttributeValue1 = "hi";
+
+    if(fs.getAbfsStore().getPrefixMode() == PrefixMode.BLOB) {
+      Assume.assumeTrue(!getIsNamespaceEnabled(fs));
+      attributeValue1 = "hi".getBytes(StandardCharsets.UTF_8);
+    }
+    else {
+      attributeValue1 = fs.getAbfsStore().encodeAttribute("hi");
+    }
+
+    // Attribute not present initially
+    assertNull(fs.getXAttr(path, attributeName1));
+
+    // Set the Attributes Multiple times
+    fs.setXAttr(path, attributeName1, attributeValue1);
```

Comment:
You may please mention that Filesystem internally adds flag: `EnumSet.of(XAttrSetFlag.CREATE       . XAttrSetFlag.REPLACE)`

```
  public void setXAttr(Path path, String name, byte[] value)
      throws IOException {
    setXAttr(path, name, value, EnumSet.of(XAttrSetFlag.CREATE,
        XAttrSetFlag.REPLACE));
  }
```
Good test!

Reply:
Added

Reply:
seems some commit is missing in the pr.

Reply:
Pushed